### PR TITLE
Imod6 1032 add 2D ponding-exchange

### PR DESCRIPTION
**Pull request by HendrikKok**

### CODE REVIEW

File: imod_coupler/drivers/dfm_metamod/msw_wrapper.py [hunk 1]
```diff
@@ -88,53 +88,33 @@ def get_surfacewater_ponding_allocation(self) -> NDArray[np.float_]:
         """
         return self.get_value_ptr("ts2dfmput")
 
-    def set_surfacewater_ponding_allocation(
-        self, ponding_allocation: NDArray[np.float_]
-    ) -> None:
-        """sets ponding volume allocation in metaswap
+    def get_surfacewater_ponding_realised(self) -> NDArray[np.float_]:
+        """gets ponding volume allocation pointer in metaSWAP
```
Comment (luitjansl) [inline imod_coupler/drivers/dfm_metamod/msw_wrapper.py:92]: what is "ponding realised" ?  is it a water height? a flux? how is it different form " ponding" (so without the realised specifier?)

File: imod_coupler/drivers/dfm_metamod/dfm_metamod.py [hunk 2]
```diff
@@ -148,38 +159,49 @@ def update(self) -> None:
...
 
-            # exchange water balance to dlfow
-            self.exchange_balans.sum_demand()
-            self.exchange_balans_2dflow(self.exchange_balans.demand["sum"])
+            # exchange ponding msw to water balance 2d
+            self.exchange_ponding_msw2dflow2d()
+
+            # exchange water balance 1d to dlfow 1d
+            self.exchange_balans_1d.sum_demand()
+            self.exchange_balans_2dfm(self.exchange_balans_1d.demand["sum"])
```
Comment (luitjansl) [inline imod_coupler/drivers/dfm_metamod/dfm_metamod.py:200]: this function name is confusing. it has "2d" in the name, but I suspect it should be interpreted as to_dfm, not someting using the 2d nodes.

File: imod_coupler/drivers/dfm_metamod/dfm_metamod.py [hunk 3]
```diff
@@ -148,38 +159,49 @@ def update(self) -> None:
...
+            # flux from metaswap sprinkling to water balance 1d
             self.exchange_sprinkling_msw2dflow1d()
 
-            # exchange water balance to dlfow
-            self.exchange_balans.sum_demand()
-            self.exchange_balans_2dflow(self.exchange_balans.demand["sum"])
+            # exchange ponding msw to water balance 2d
+            self.exchange_ponding_msw2dflow2d()
+
+            # exchange water balance 1d to dlfow 1d
```
Comment (luitjansl) [inline imod_coupler/drivers/dfm_metamod/dfm_metamod.py:198]: should'nt this comment be 2 lines lower?

File: imod_coupler/drivers/dfm_metamod/dfm_metamod.py [hunk 4]
```diff
@@ -148,38 +159,49 @@ def update(self) -> None:
...
+            self.exchange_balans_1d.sum_demand()
+            self.exchange_balans_2dfm(self.exchange_balans_1d.demand["sum"])
 
             # get cummelative flux before dfm-run
-            q_dflow0 = self.dfm.get_cumulative_fluxes_1d_nodes()
+            q_dflow0_1d = self.dfm.get_cumulative_fluxes_1d_nodes()
```
Comment (luitjansl) [inline imod_coupler/drivers/dfm_metamod/dfm_metamod.py:203]: feel free to use longer function names. maybe q_dflow_1d_before_solve or something like that

File: imod_coupler/drivers/dfm_metamod/dfm_metamod.py [hunk 5]
```diff
@@ -335,6 +369,73 @@ def exchange_stage_1d_dfm2mf6(self) -> None:
             updated_river_stage,
         )
 
+    def exchange_stage_2d_dfm2msw(self) -> None:
```
Comment (luitjansl) [inline imod_coupler/drivers/dfm_metamod/dfm_metamod.py:372]: can we make a unittest for this function?

File: imod_coupler/drivers/dfm_metamod/mapping.py [hunk 6]
```diff
@@ -531,6 +532,48 @@ def get_dflow1d_lookup(self) -> None:
...
+        of dflow, dict is filled with node-id's corresponding tot xy-pairs.
+        this functions should be called after initialisation of dflow-fm.
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        tuple[dict[tuple[float, float], int], bool]
```
Comment (luitjansl) [inline imod_coupler/drivers/dfm_metamod/mapping.py:547]: it doesn't return anything

**Review by luitjansl — COMMENTED** [review]
nice work! I added some minor comments. But I think some tests should be added for the new exchange class.

File: imod_coupler/drivers/dfm_metamod/msw_wrapper.py (hunk 1 above)
Comment (HendrikKok) [reply imod_coupler/drivers/dfm_metamod/msw_wrapper.py:92]: Its a water volume per dtsw-timestep. MetaSWAP uses 2 different variables for allocation and realization phase of ponding water. So we have to communicate with two different pointers unfortunately.

File: imod_coupler/drivers/dfm_metamod/dfm_metamod.py (hunk 2 above)
Comment (HendrikKok) [reply imod_coupler/drivers/dfm_metamod/dfm_metamod.py:200]: yes, its confusing. I think we have it at different locations. maybe a refactor idea?

File: imod_coupler/drivers/dfm_metamod/dfm_metamod.py (hunk 4 above)
Comment (HendrikKok) [reply imod_coupler/drivers/dfm_metamod/dfm_metamod.py:203]: done

File: imod_coupler/drivers/dfm_metamod/dfm_metamod.py (hunk 3 above)
Comment (HendrikKok) [reply imod_coupler/drivers/dfm_metamod/dfm_metamod.py:198]: dont know what you mean?

Comment (luitjansl) [reply imod_coupler/drivers/dfm_metamod/dfm_metamod.py:198]: just a detail. the comment "exchange water balance 1d to dlfow 1d" is just above the code-line  self.exchange_balans_1d.sum_demand()
but the actual exchange to dflow is done 2 lines below, on the code-line self.exchange_balans_2dfm(self.exchange_balans_1d.demand["sum"]), i believe. thats all.

File: imod_coupler/drivers/dfm_metamod/msw_wrapper.py (hunk 1 above)
Comment (luitjansl) [reply imod_coupler/drivers/dfm_metamod/msw_wrapper.py:92]: can you add a comment in the code with this text? it would help

**Review by luitjansl — APPROVED** [review]

File: imod_coupler/drivers/dfm_metamod/dfm_metamod.py (hunk 5 above)
Comment (HendrikKok) [reply imod_coupler/drivers/dfm_metamod/dfm_metamod.py:376]: created issue IMOD6-1051 for this

**Pull request merged**